File: event_guests.rb
# event_guests.rb
#!/usr/bin/env ruby
require 'json'
require 'securerandom'
require 'date'
require 'csv'

DATA_FILE = 'events.json'

class Guest
  attr_accessor :id, :name, :email, :rsvp, :plus_one, :dietary, :notes, :added

  def initialize(name, email, plus_one = false, dietary = '', notes = '')
    @id = SecureRandom.hex(4)
    @name = name
    @email = email
    @rsvp = 'Pending'
    @plus_one = plus_one
    @dietary = dietary
    @notes = notes
    @added = Time.now.iso8601
  end

  def to_hash
    { id: @id, name: @name, email: @email, rsvp: @rsvp,
      plus_one: @plus_one, dietary: @dietary, notes: @notes, added: @added }
  end

  def self.from_hash(h)
    g = new(h['name'], h['email'], h['plus_one'], h['dietary'], h['notes'])
    g.id = h['id']
    g.rsvp = h['rsvp'] || 'Pending'
    g.added = h['added'] || Time.now.iso8601
    g
  end
end

class Event
  attr_accessor :id, :name, :date, :venue, :guests, :created

  def initialize(name, date = '', venue = '')
    @id = SecureRandom.hex(4)
    @name = name
    @date = date
    @venue = venue
    @guests = []
    @created = Time.now.iso8601
  end

  def to_hash
    { id: @id, name: @name, date: @date, venue: @venue,
      guests: @guests.map(&:to_hash), created: @created }
  end

  def self.from_hash(h)
    e = new(h['name'], h['date'], h['venue'])
    e.id = h['id']
    e.guests = h['guests'].map { |g| Guest.from_hash(g) }
    e.created = h['created'] || Time.now.iso8601
    e
  end
end

class Organizer
  attr_reader :events

  def initialize
    @events = []
    load
  end

  def load
    if File.exist?(DATA_FILE)
      data = JSON.parse(File.read(DATA_FILE))
      @events = data.map { |h| Event.from_hash(h) }
    end
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate(@events.map(&:to_hash)))
  end

  def get_event(id)
    @events.find { |e| e.id == id }
  end

  def get_guest(event, id)
    event.guests.find { |g| g.id == id }
  end

  def create(name, date = '', venue = '')
    e = Event.new(name, date, venue)
    @events << e
    save
    puts "✅ Event created: #{e.name} (ID: #{e.id})"
  end

  def add_guest(event_id, name, email, plus_one = false, dietary = '', notes = '')
    e = get_event(event_id)
    unless e
      puts "Event #{event_id} not found."
      return
    end
    g = Guest.new(name, email, plus_one, dietary, notes)
    e.guests << g
    save
    puts "✅ Guest added: #{g.name} (ID: #{g.id})"
  end

  def list_guests(event_id, status = nil)
    e = get_event(event_id)
    unless e
      puts "Event #{event_id} not found."
      return
    end
    guests = e.guests
    guests = guests.select { |g| g.rsvp == status } if status
    if guests.empty?
      puts "No guests."
      return
    end
    puts "\n📋 Event: #{e.name}"
    puts "Date: #{e.date}" unless e.date.empty?
    puts "Venue: #{e.venue}" unless e.venue.empty?
    puts "\n👤 Guests (#{guests.size}):"
    guests.each_with_index do |g, i|
      plus = g.plus_one ? ' (plus‑one: Yes)' : ' (plus‑one: No)'
      dietary = g.dietary.empty? ? '' : " 🥗 #{g.dietary}"
      notes = g.notes.empty? ? '' : " 📝 #{g.notes}"
      puts "  #{i+1}. #{g.name} (#{g.email}) – #{g.rsvp}#{plus}#{dietary}#{notes}"
    end
  end

  def rsvp(event_id, guest_id, status)
    e = get_event(event_id)
    unless e
      puts "Event #{event_id} not found."
      return
    end
    g = get_guest(e, guest_id)
    unless g
      puts "Guest #{guest_id} not found."
      return
    end
    valid_statuses = ['Pending', 'Attending', 'Declined']
    unless valid_statuses.include?(status)
      puts "Invalid status. Choose: #{valid_statuses.join(', ')}"
      return
    end
    g.rsvp = status
    save
    puts "✅ #{g.name} RSVP updated to: #{status}"
  end

  def stats(event_id)
    e = get_event(event_id)
    unless e
      puts "Event #{event_id} not found."
      return
    end
    total = e.guests.size
    attending = e.guests.count { |g| g.rsvp == 'Attending' }
    pending = e.guests.count { |g| g.rsvp == 'Pending' }
    declined = e.guests.count { |g| g.rsvp == 'Declined' }
    plus_ones = e.guests.count(&:plus_one)
    puts "\n📊 Event: #{e.name}"
    puts "  Total guests: #{total}"
    puts "  Attending: #{attending}"
    puts "  Pending: #{pending}"
    puts "  Declined: #{declined}"
    puts "  Plus‑ones: #{plus_ones}"
  end

  def search(event_id, term)
    e = get_event(event_id)
    unless e
      puts "Event #{event_id} not found."
      return
    end
    lower = term.downcase
    results = e.guests.select { |g| g.name.downcase.include?(lower) || g.email.downcase.include?(lower) }
    if results.empty?
      puts "No matches."
      return
    end
    puts "\n🔍 Found #{results.size} guest(s):"
    results.each_with_index do |g, i|
      puts "  #{i+1}. #{g.name} (#{g.email}) – #{g.rsvp}"
    end
  end

  def export(event_id, filename)
    e = get_event(event_id)
    unless e
      puts "Event #{event_id} not found."
      return
    end
    CSV.open(filename, 'w') do |csv|
      csv << ['ID', 'Name', 'Email', 'RSVP', 'Plus-One', 'Dietary', 'Notes', 'Added']
      e.guests.each do |g|
        csv << [g.id, g.name, g.email, g.rsvp, g.plus_one, g.dietary, g.notes, g.added]
      end
    end
    puts "✅ Exported #{e.guests.size} guests to #{filename}"
  end
end

if ARGV.empty?
  puts "Usage: event_guests.rb <command> [options]"
  exit
end

app = Organizer.new
cmd = ARGV.shift

case cmd
when 'create'
  name = ARGV.shift
  if name.nil?
    puts "create <name> [--date DATE] [--venue VENUE]"
    exit
  end
  date = ''
  venue = ''
  while ARGV.any?
    case ARGV[0]
    when '--date'
      ARGV.shift
      date = ARGV.shift || ''
    when '--venue'
      ARGV.shift
      venue = ARGV.shift || ''
    else
      break
    end
  end
  app.create(name, date, venue)

when 'add-guest'
  if ARGV.size < 3
    puts "add-guest <event_id> <name> <email> [--plus-one] [--dietary DIET] [--notes NOTES]"
    exit
  end
  event_id = ARGV.shift
  name = ARGV.shift
  email = ARGV.shift
  plus_one = false
  dietary = ''
  notes = ''
  while ARGV.any?
    case ARGV[0]
    when '--plus-one'
      ARGV.shift
      plus_one = true
    when '--dietary'
      ARGV.shift
      dietary = ARGV.shift || ''
    when '--notes'
      ARGV.shift
      notes = ARGV.shift || ''
    else
      break
    end
  end
  app.add_guest(event_id, name, email, plus_one, dietary, notes)

when 'list-guests'
  event_id = ARGV.shift
  if event_id.nil?
    puts "list-guests <event_id> [--status STATUS]"
    exit
  end
  status = nil
  if ARGV.include?('--status')
    idx = ARGV.index('--status')
    status = ARGV[idx+1] if idx
  end
  app.list_guests(event_id, status)

when 'rsvp'
  if ARGV.size < 3
    puts "rsvp <event_id> <guest_id> <status>"
    exit
  end
  event_id = ARGV.shift
  guest_id = ARGV.shift
  status = ARGV.shift
  app.rsvp(event_id, guest_id, status)

when 'stats'
  event_id = ARGV.shift
  if event_id.nil?
    puts "stats <event_id>"
    exit
  end
  app.stats(event_id)

when 'search'
  if ARGV.size < 2
    puts "search <event_id> <term>"
    exit
  end
  event_id = ARGV.shift
  term = ARGV.shift
  app.search(event_id, term)

when 'export'
  event_id = ARGV.shift
  if event_id.nil?
    puts "export <event_id> [--filename FILE]"
    exit
  end
  filename = 'guests.csv'
  if ARGV.include?('--filename')
    idx = ARGV.index('--filename')
    filename = ARGV[idx+1] if idx
  end
  app.export(event_id, filename)

else
  puts "Unknown command. Use create, add-guest, list-guests, rsvp, stats, search, export."
end

File: event_guests.js
// event_guests.js
#!/usr/bin/env node
const fs = require('fs');
const { program } = require('commander');
const { v4: uuidv4 } = require('uuid');

const DATA_FILE = 'events.json';

class Guest {
    constructor(name, email, plusOne = false, dietary = '', notes = '') {
        this.id = uuidv4().slice(0,8);
        this.name = name;
        this.email = email;
        this.rsvp = 'Pending';
        this.plusOne = plusOne;
        this.dietary = dietary;
        this.notes = notes;
        this.added = new Date().toISOString();
    }
}

class Event {
    constructor(name, date = '', venue = '') {
        this.id = uuidv4().slice(0,8);
        this.name = name;
        this.date = date;
        this.venue = venue;
        this.guests = [];
        this.created = new Date().toISOString();
    }
}

class Organizer {
    constructor() {
        this.events = [];
        this.load();
    }

    load() {
        if (fs.existsSync(DATA_FILE)) {
            try {
                const data = JSON.parse(fs.readFileSync(DATA_FILE));
                this.events = data.map(e => {
                    const ev = new Event(e.name, e.date, e.venue);
                    ev.id = e.id;
                    ev.guests = e.guests.map(g => {
                        const guest = new Guest(g.name, g.email, g.plusOne, g.dietary, g.notes);
                        guest.id = g.id;
                        guest.rsvp = g.rsvp || 'Pending';
                        guest.added = g.added;
                        return guest;
                    });
                    ev.created = e.created;
                    return ev;
                });
            } catch (e) {}
        }
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify(this.events, null, 2));
    }

    getEvent(id) {
        return this.events.find(e => e.id === id);
    }

    getGuest(event, id) {
        return event.guests.find(g => g.id === id);
    }

    create(name, date, venue) {
        const e = new Event(name, date, venue);
        this.events.push(e);
        this.save();
        console.log(`✅ Event created: ${e.name} (ID: ${e.id})`);
    }

    addGuest(eventId, name, email, plusOne, dietary, notes) {
        const e = this.getEvent(eventId);
        if (!e) {
            console.log(`Event ${eventId} not found.`);
            return;
        }
        const g = new Guest(name, email, plusOne, dietary, notes);
        e.guests.push(g);
        this.save();
        console.log(`✅ Guest added: ${g.name} (ID: ${g.id})`);
    }

    listGuests(eventId, status) {
        const e = this.getEvent(eventId);
        if (!e) {
            console.log(`Event ${eventId} not found.`);
            return;
        }
        let guests = e.guests;
        if (status) {
            guests = guests.filter(g => g.rsvp === status);
        }
        if (guests.length === 0) {
            console.log('No guests.');
            return;
        }
        console.log(`\n📋 Event: ${e.name}`);
        if (e.date) console.log(`Date: ${e.date}`);
        if (e.venue) console.log(`Venue: ${e.venue}`);
        console.log(`\n👤 Guests (${guests.length}):`);
        guests.forEach((g, i) => {
            const plus = g.plusOne ? ' (plus‑one: Yes)' : ' (plus‑one: No)';
            const dietary = g.dietary ? ` 🥗 ${g.dietary}` : '';
            const notes = g.notes ? ` 📝 ${g.notes}` : '';
            console.log(`  ${i+1}. ${g.name} (${g.email}) – ${g.rsvp}${plus}${dietary}${notes}`);
        });
    }

    rsvp(eventId, guestId, status) {
        const e = this.getEvent(eventId);
        if (!e) {
            console.log(`Event ${eventId} not found.`);
            return;
        }
        const g = this.getGuest(e, guestId);
        if (!g) {
            console.log(`Guest ${guestId} not found.`);
            return;
        }
        const validStatuses = ['Pending', 'Attending', 'Declined'];
        if (!validStatuses.includes(status)) {
            console.log(`Invalid status. Choose: ${validStatuses.join(', ')}`);
            return;
        }
        g.rsvp = status;
        this.save();
        console.log(`✅ ${g.name} RSVP updated to: ${status}`);
    }

    stats(eventId) {
        const e = this.getEvent(eventId);
        if (!e) {
            console.log(`Event ${eventId} not found.`);
            return;
        }
        const total = e.guests.length;
        const attending = e.guests.filter(g => g.rsvp === 'Attending').length;
        const pending = e.guests.filter(g => g.rsvp === 'Pending').length;
        const declined = e.guests.filter(g => g.rsvp === 'Declined').length;
        const plusOnes = e.guests.filter(g => g.plusOne).length;
        console.log(`\n📊 Event: ${e.name}`);
        console.log(`  Total guests: ${total}`);
        console.log(`  Attending: ${attending}`);
        console.log(`  Pending: ${pending}`);
        console.log(`  Declined: ${declined}`);
        console.log(`  Plus‑ones: ${plusOnes}`);
    }

    search(eventId, term) {
        const e = this.getEvent(eventId);
        if (!e) {
            console.log(`Event ${eventId} not found.`);
            return;
        }
        const lower = term.toLowerCase();
        const results = e.guests.filter(g =>
            g.name.toLowerCase().includes(lower) ||
            g.email.toLowerCase().includes(lower)
        );
        if (results.length === 0) {
            console.log('No matches.');
            return;
        }
        console.log(`\n🔍 Found ${results.length} guest(s):`);
        results.forEach((g, i) => {
            console.log(`  ${i+1}. ${g.name} (${g.email}) – ${g.rsvp}`);
        });
    }

    export(eventId, filename) {
        const e = this.getEvent(eventId);
        if (!e) {
            console.log(`Event ${eventId} not found.`);
            return;
        }
        let csv = 'ID,Name,Email,RSVP,Plus-One,Dietary,Notes,Added\n';
        for (const g of e.guests) {
            csv += `${g.id},${g.name},${g.email},${g.rsvp},${g.plusOne},${g.dietary},${g.notes},${g.added}\n`;
        }
        fs.writeFileSync(filename, csv);
        console.log(`✅ Exported ${e.guests.length} guests to ${filename}`);
    }
}

program
    .command('create <name>')
    .option('--date <date>', 'Event date')
    .option('--venue <venue>', 'Event venue')
    .action((name, options) => {
        const app = new Organizer();
        app.create(name, options.date || '', options.venue || '');
    });

program
    .command('add-guest <eventId> <name> <email>')
    .option('--plus-one', 'Guest can bring a plus-one')
    .option('--dietary <diet>', 'Dietary restrictions')
    .option('--notes <notes>', 'Additional notes')
    .action((eventId, name, email, options) => {
        const app = new Organizer();
        app.addGuest(eventId, name, email, options.plusOne || false, options.dietary || '', options.notes || '');
    });

program
    .command('list-guests <eventId>')
    .option('--status <status>', 'Filter by status (Pending, Attending, Declined)')
    .action((eventId, options) => {
        const app = new Organizer();
        app.listGuests(eventId, options.status || '');
    });

program
    .command('rsvp <eventId> <guestId> <status>')
    .action((eventId, guestId, status) => {
        const app = new Organizer();
        app.rsvp(eventId, guestId, status);
    });

program
    .command('stats <eventId>')
    .action((eventId) => {
        const app = new Organizer();
        app.stats(eventId);
    });

program
    .command('search <eventId> <term>')
    .action((eventId, term) => {
        const app = new Organizer();
        app.search(eventId, term);
    });

program
    .command('export <eventId>')
    .option('--filename <filename>', 'Output filename', 'guests.csv')
    .action((eventId, options) => {
        const app = new Organizer();
        app.export(eventId, options.filename);
    });

program.parse(process.argv);
